🐛 BUG: Uncaught TypeError: Class extends value undefined is not a constructor or null

### Which Cloudflare product(s) does this pertain to?

Wrangler core, Other

### What version(s) of the tool(s) are you using?

Wrangler 3.5.0

### What version of Node are you using?

18.17.0

### What operating system are you using?

Mac

### Describe the Bug

When creating a worker that defines a Durable Object (but doesn't export it in the main file), a very cryptic error is thrown
```
service core:user:my-worker: Uncaught TypeError: Class extends value undefined is not a constructor or null
  at index.js:1648:44 in maskDurableObjectDefinition
  at index.js:1657:26
✘ [ERROR] MiniflareCoreError [ERR_RUNTIME_FAILURE]: The Workers runtime failed to start. There is likely additional logging output above.
```

We should probably check whether the main script exports the DO, and if it doesn't, provide an easier to understand error. 

### Please provide a link to a minimal reproduction

```ts
export default {}
// note, the lack of `export class MyDO`
```
```toml
name = "my-worker"
script = "src/index.ts"
compatibility_date = "2023-08-11"

[[durable_objects.bindings]]
name = "MY_DO"
class_name = "MyDO"

[[migrations]]
tag = "v1"
new_classes = ["MyDo"]
```

### Please provide any relevant error logs

_No response_

--- a/.changeset/rude-eagles-buy.md
+++ b/.changeset/rude-eagles-buy.md
@@ -1,0 +1,7 @@
+---
+"wrangler": patch
+---
+
+Improve the error message when a script isn't exported a Durable Object class
+
+Previously, wrangler would error with a message like `Uncaught TypeError: Class extends value undefined is not a constructor or null`. This improves that messaging to be more understandable to users.

--- a/fixtures/durable-objects-app/package.json
+++ b/fixtures/durable-objects-app/package.json
@@ -1,0 +1,20 @@
+{
+	"name": "do-worker-app",
+	"version": "1.0.0",
+	"private": true,
+	"description": "",
+	"license": "ISC",
+	"author": "",
+	"main": "src/index.js",
+	"scripts": {
+		"dev": "wrangler deploy --dry-run",
+		"test": "npx vitest run",
+		"test:ci": "npx vitest run",
+		"test:watch": "npx vitest"
+	},
+	"devDependencies": {
+		"@cloudflare/workers-tsconfig": "workspace:^",
+		"undici": "^5.23.0",
+		"wrangler": "workspace:*"
+	}
+}

--- a/fixtures/durable-objects-app/src/index.js
+++ b/fixtures/durable-objects-app/src/index.js
@@ -1,0 +1,5 @@
+export default {
+	async fetch() {
+		return new Response("foo");
+	},
+};

--- a/fixtures/durable-objects-app/tests/index.test.ts
+++ b/fixtures/durable-objects-app/tests/index.test.ts
@@ -1,0 +1,29 @@
+import { resolve } from "node:path";
+import { assert, describe, it } from "vitest";
+import { execSync } from "node:child_process";
+
+describe("durable objects", () => {
+	it("should throw an error when the worker doesn't export a durable object but requires one", ({
+		expect,
+	}) => {
+		let err: string = "";
+		try {
+			execSync("pnpm run dev", {
+				cwd: resolve(__dirname, ".."),
+			});
+			assert(false); // Should never reach this
+		} catch (e) {
+			err = (e as Error).message
+				.replaceAll("✘", "X")
+				.replace(/\\/g, "/")
+				.replace(/[^\S\n]+\n/g, "\n")
+				.trimEnd();
+		}
+		expect(err).toMatchInlineSnapshot(`
+			"Command failed: pnpm run dev
+			[31mX [41;31m[[41;97mERROR[41;31m][0m [1mYour Worker depends on the following Durable Objects, which are not exported in your entrypoint file: FooBar.[0m
+
+			  You should export these objects from your entrypoint, src/index.js."
+		`);
+	});
+});

--- a/fixtures/durable-objects-app/tests/tsconfig.json
+++ b/fixtures/durable-objects-app/tests/tsconfig.json
@@ -1,0 +1,7 @@
+{
+	"extends": "@cloudflare/workers-tsconfig/tsconfig.json",
+	"compilerOptions": {
+		"types": ["node"]
+	},
+	"include": ["**/*.ts", "../../../node-types.d.ts"]
+}

--- a/fixtures/durable-objects-app/tsconfig.json
+++ b/fixtures/durable-objects-app/tsconfig.json
@@ -1,0 +1,12 @@
+{
+	"compilerOptions": {
+		"target": "ES2020",
+		"esModuleInterop": true,
+		"module": "CommonJS",
+		"lib": ["ES2020"],
+		"types": ["node"],
+		"moduleResolution": "node",
+		"noEmit": true
+	},
+	"include": ["tests", "../../node-types.d.ts"]
+}

--- a/fixtures/durable-objects-app/wrangler.toml
+++ b/fixtures/durable-objects-app/wrangler.toml
@@ -1,0 +1,13 @@
+name = "do-worker-app"
+compatibility_date = "2022-03-31"
+
+main = "src/index.js"
+
+[durable_objects]
+bindings = [
+	{ name = "MY_DO", class_name = "FooBar" }
+]
+
+[[migrations]]
+tag = "v1"
+new_classes = ["FooBar"]

--- a/packages/wrangler/src/__tests__/deploy.test.ts
+++ b/packages/wrangler/src/__tests__/deploy.test.ts
@@ -7654,7 +7654,15 @@ export default{
 				},
 				migrations: [{ tag: "v1", new_classes: ["SomeClass"] }],
 			});
-			writeWorkerSource();
+			fs.writeFileSync(
+				"index.js",
+				`export default {
+        	async fetch(request) {
+          	return new Response('Hello' + foo);
+        	},
+      	};
+				export class SomeClass {};`
+			);
 			process.env.CLOUDFLARE_ACCOUNT_ID = "";
 			await runWrangler("deploy index.js --dry-run");
 			expect(std).toMatchInlineSnapshot(`

--- a/packages/wrangler/src/deployment-bundle/bundle.ts
+++ b/packages/wrangler/src/deployment-bundle/bundle.ts
@@ -378,6 +378,18 @@ export async function bundleWorker(
 	}
 
 	const entryPoint = getEntryPointFromMetafile(entryFile, result.metafile);
+	const notExportedDOs = doBindings
+		.filter((x) => !x.script_name && !entryPoint.exports.includes(x.class_name))
+		.map((x) => x.class_name);
+	if (notExportedDOs.length) {
+		const relativePath = path.relative(process.cwd(), entryFile);
+		throw new Error(
+			`Your Worker depends on the following Durable Objects, which are not exported in your entrypoint file: ${notExportedDOs.join(
+				", "
+			)}.\nYou should export these objects from your entrypoint, ${relativePath}.`
+		);
+	}
+
 	const bundleType = entryPoint.exports.length > 0 ? "esm" : "commonjs";
 
 	const sourceMapPath = Object.keys(result.metafile.outputs).filter((_path) =>

